metricbeat/mb/lightmodules.go

```diff
@@ -0,0 +1,288 @@
+			m.Module, m.Name,
+			m.Input.Module, m.Input.MetricSet)
+	}
+
+	originalFactory := registration.Factory
+	registration.IsDefault = m.Default
+
+	// Light modules factory has to override defaults and reproduce builder
+	// functionality with the resulting configuration, it does:
+	// - Override defaults
+	// - Call module factory if registered (it wouldn't have been called
+	//   if light module is really a registered mixed module)
+	// - Call host parser if defined (it would have already been called
+	//   without the light module defaults)
+	// - Finally, call the original factory for the registered metricset
+	registration.Factory = func(base BaseMetricSet) (MetricSet, error) {
+		// Override default config on base module and metricset
+		base.name = m.Name
+		baseModule, err := m.baseModule(base.module)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create base module for light module")
+		}
+		base.module = baseModule
+
+		// Run module factory if registered, it will be called once per
+		// metricset, but it should be idempotent
+		moduleFactory := r.moduleFactory(m.Input.Module)
+		if moduleFactory != nil {
+			m, err := moduleFactory(*baseModule)
+			if err != nil {
+				return nil, err
+			}
+			base.module = m
+		}
+
+		// At this point host parser was already run, we need to run this again
+		// with the overriden defaults
+		if registration.HostParser != nil {
+			base.hostData, err = registration.HostParser(base.module, base.host)
+			if err != nil {
+				return nil, err
+			}
+			base.host = base.hostData.Host
+		}
+
+		return originalFactory(base)
+	}
+
+	return registration, nil
+}
+
+// baseModule does the configuration overrides in the base module configuration
+// taking into account the light metric set default configurations
+func (m *LightMetricSet) baseModule(from Module) (*BaseModule, error) {
+	baseModule := BaseModule{
+		name:   m.Module,
+		config: from.Config(),
+	}
+	var err error
+	// Set defaults
+	if baseModule.rawConfig, err = common.NewConfigFrom(m.Input.Defaults); err != nil {
+		return nil, errors.Wrap(err, "invalid input defaults")
+	}
+	// Copy values from user configuration
+	if err = from.UnpackConfig(baseModule.rawConfig); err != nil {
+		return nil, errors.Wrap(err, "failed to copy values from user configuration")
+	}
+	// Update module configuration
+	if err = baseModule.UnpackConfig(&baseModule.config); err != nil {
+		return nil, errors.Wrap(err, "failed to set module configuration")
+	}
+	return &baseModule, nil
+}
+
+func (r *LightModulesRegistry) readModule(moduleName string) (*LightModule, bool, error) {
+	modulePath, found := r.findModulePath(moduleName)
+	if !found {
+		return nil, false, nil
+	}
+
+	moduleConfig, err := r.readModuleConfig(modulePath)
+	if err != nil {
+		return nil, true, errors.Wrapf(err, "failed to load light module '%s' definition", moduleName)
+	}
+
+	metricSets := make(map[string]LightMetricSet)
```

Comment:
Add this bunch of code to its own method so we won't have to follow the loop to quickly understand what it is doing.

In other words, when I see `readModule` method I want a less than 10-15 loc method which _finds the module path_, _reads the module (open it)_ and... _read its metricsets_?

BTW, _Read_ is an action that usually returns streams of data. I think it's more widely accepted to use `Open` in many of this cases. In Go, for example, you open a file with `os.Open(string)(*File, error)` and returns a struct. In contrast, you can also read it using its method `Read`, which implements `io.Reader` and works byte by byte.

It's just an example, don't use _Open_ if you don't like it, but I think that _read_ is really misleading to follow the rest of the code. :)